scripts/volmount/crypt

```diff
@@ -0,0 +1,143 @@
+#!/bin/sh
+
+set -e
+
+op=$1
+dm_type=$2
+path=$3
+mntpath=$4
+
+CAAM_KEYGEN=/lib/pv/volmount/caam-keygen
+
+do_find_free_loop()
+{
+	for j in `ls /dev/loop[0-9]*`
+	do
+		free=0
+		for i in `losetup -a | awk '{split($0,a,":"); print a[1]}'`
+		do
+			if [ $j == $i ]
+			then
+				free=1
+				break
+			fi
+		done
+		[ $free -eq 0 ] && break
+	done
+	FREE_LOOP=$j
+}
+
+do_crypt_init()
+{
+	base=`basename $1`
+	dir=`dirname $1`
+
+	mkdir -p $dir
+
+	keyname=`echo $base | awk '{split($0,a,","); print a[4]}'`
+	$CAAM_KEYGEN create $keyname ecb -s 16
+	cp /data/caam/$keyname* $dir
+	cat /data/caam/$keyname | /bin/keyctl padd logon logkey: @u
+
+	img_file=`echo $base | awk '{split($0,a,","); print a[1]}'`
+	img_size=`echo $base | awk '{split($0,a,","); print a[2]}'`
+	/bin/dd if=/dev/zero of=$dir/$img_file bs=1M count=$img_size
+	do_find_free_loop
+
+	/sbin/losetup $FREE_LOOP $dir/$img_file
+	[ $? -ne 0 ] && exit 1
+
+	device_name=`echo $img_file | awk '{split($0,a,"."); print a[1]}'`
+	/usr/sbin/dmsetup -v create $device_name --table "0 $(/sbin/blockdev --getsz $FREE_LOOP) crypt capi:tk(cbc(aes))-plain :36:logon:logkey: 0 $FREE_LOOP 0 1 sector_size:512"
+	[ ! -b /dev/mapper/$device_name ] && exit 1
+
+	/usr/sbin/mkfs.ext4 /dev/mapper/$device_name
+
+	/usr/sbin/dmsetup remove $device_name
+	/sbin/losetup -d $FREE_LOOP
+	/bin/keyctl clear @u
+	/bin/keyctl purge @u
+	rm -rf /data
+}
+
+do_dm_sanity()
+{
+	[ ! -f $CAAM_KEYGEN ] && exit 1
+	[ ! -f /bin/keyctl ] && exit 1
+	[ ! -f /lib/pv/volmount/caam-keygen ] && exit 1
+	[ ! -f /sbin/losetup ] && exit 1
+	[ ! -f /usr/sbin/dmsetup ] && exit 1
+	[ ! -f /bin/dd ] && exit 1
+	[ ! -f /usr/sbin/mkfs.ext4 ] && exit 1
+
+	export LD_LIBRARY_PATH=/x86_64-linux-gnu
+}
+
+do_mount_disk()
+{
+	do_dm_sanity
+
+	base=`basename $2`
+	dir=`dirname $2`
+	keyname=`echo $base | awk '{split($0,a,","); print a[4]}'`
+	img_file=`echo $base | awk '{split($0,a,","); print a[1]}'`
+	device_name=`echo $img_file | awk '{split($0,a,"."); print a[1]}'`
+	if [ -b /dev/mapper/$device_name ]
+	then
+		mkdir -p $3
+		mount /dev/mapper/$device_name $3
+		[ $? -ne 0 ] && exit 1
+
+		return
+	fi
+
+	[ ! -d $dir ] && do_crypt_init $2
+
+	cd $dir
+	mkdir -p /data/caam
+	cp $keyname* /data/caam
+	$CAAM_KEYGEN import $keyname.bb $keyname.import
+
+	cat /data/caam/$keyname.import | /bin/keyctl padd logon logkey: @u
+
+	do_find_free_loop
+	/sbin/losetup $FREE_LOOP $dir/$img_file
+	[ $? -ne 0 ] && exit 1
```

Comment:
I guess it would be interesting to print some error message and return different error codes like @asac is doing in https://github.com/pantavisor/pantavisor/pull/19. Then, you could use tsh_run_output to get that, which is already fixed in master I think

Reply:
added debug and consumed with tsh_run_output